@here/harp-datasource-protocol/lib/Expr.ts

```diff
@@ -87,6 +89,38 @@ export abstract class Expr {
                 });
                 return new ContainsExpr(this.fromJSON(node[1]), elements);
 
+            case "match": {
+                const value = this.fromJSON(node[1]);
+                const conditions: Array<[MatchLabel, Expr]> = [];
+                for (let i = 2; i < node.length - 1; i += 2) {
+                    const label = node[i];
+                    if (
+                        !(
+                            typeof label === "number" ||
+                            typeof label === "string" ||
+                            Array.isArray(label)
+                        )
+                    ) {
+                        throw new Error(`parse error ${JSON.stringify(label)}`);
+                    }
+                    const expr = this.fromJSON(node[i + 1]);
+                    conditions.push([label, expr]);
+                }
+                const fallback = this.fromJSON(node[node.length - 1]);
+                return new MatchExpr(value, conditions, fallback);
+            }
+
+            case "case": {
+                const branches: Array<[Expr, Expr]> = [];
+                for (let i = 1; i < node.length - 1; i += 2) {
+                    const condition = this.fromJSON(node[i]);
+                    const expr = this.fromJSON(node[i + 1]);
+                    branches.push([condition, expr]);
+                }
+                const caseFallback = this.fromJSON(node[node.length - 1]);
```

Comment:
Some error handling is needed.
In case of missing `fallback` element, expr from last pair will whereas `null/undefined` would be expected.